base/dict.jl

```diff
@@ -713,13 +713,19 @@ function start(t::Dict)
     return i
 end
 done(t::Dict, i) = i > length(t.vals)
-next(t::Dict{K,V}, i) where {K,V} = (Pair{K,V}(t.keys[i],t.vals[i]), skip_deleted(t,i+1))
+function next(t::Dict{K,V}, i) where {K,V}
+    @inbounds return (Pair{K,V}(t.keys[i],t.vals[i]), skip_deleted(t,i+1))
```

Comment:
I don't think this is safe --- it's possible some user code could pass a bad value of `i` by mistake.

Reply:
These could be marked `@propagate_inbounds`, though; that seems to be what arrays do.

Reply:
Good point!

Reply:
Ref https://github.com/JuliaLang/julia/pull/15291

Reply:
I'm assuming that what is now here makes sense? And that `for` loops expand/lower to include `@inbounds next(...)` for fast iteration over e.g. `Array`?

Reply:
No, for loop cannot be lowered to `@inbounds`. It only guarantee that the iterator protocol is properly executed, but not inbounds access in `next`.

Reply:
Umm... so are you telling me that `for x in vector` isn't optimally fast? It really spends time checking bounds, immediately after it has verified `next`?

Reply:
Correct.

Reply:
Oh dear, I see #15291 wasn't merged... (I figured it was merged and we've moved onto something simpler, but alas)

Reply:
Thanks @yuyichao for clearing that up. At least with `@propagate_inbounds` here, users have the option of `@inbounds for kv in dict` for a speedup, right?

Also, I'll have to improve my benchmark.